### PR TITLE
TwentyTwentyOne: Restructure button style cascade to improve button color options.

### LINKED ISSUE #858: Button links in cover block inherit the wrong link colors
When creating a cover block with a call to action button, the `color: currentColor` style has the wrong priority and ends up overwriting the intended color.

This seems to be the same issue as with #209, and originates from styles which accounts for regular text links, but not buttons, found at at https://github.com/WordPress/twentytwentyone/blob/d4c6fecb56d7e8fbfa859129057b8d1444e03063/style.css#L2039-L2046

Example:
Cover block with a background image and a button.
The button has a blue background, and white text chosen from the available color palette, this applies the `has-white-color` class to the button link.

The result:
dark button with dark text

Expected result:
dark button with white text

### LINKED ISSUE #880: Editor: With Dark Mode enabled, buttons are only white if the selected body background is the default color
**Describe the bug**
If the user selects or keeps the mint green body background color, and enables dark mode,
the button block color is changed to white in the editor and the front.

If the user selects another light body background color from the palette or color picker, and enables dark mode,
the button block color remains dark grey in the editor, but is changed white on the front.

If Dark Mode is disabled, and the user selects a dark body background, the button block color is changed to white in the editor and the front.

**To Reproduce**
1. Enable Dark Mode
2. Change the body background color to a light palette color, but not mint.
3. _-It does not matter in which order you enable dark mode or change the color_
4. Add a default button block in the editor.
5. See that the editor background color is grey and the button is the default dark grey, not the white color.
6. Add some text in the content and confirm that the text color is white.

**Expected behavior**
With Dark Mode, the buttons colors should be adjusted to improve readability, even if the color selected in the customizer option is light.

**Screenshots**

**Editor version (please complete the following information):**
- WordPress version: 5.6-RC1-49687, with theme version from lastest trunk on GitHub
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default?  Default

### PR DESCRIPTION
**Pull request by allancole**
Fixes #858
_Possibly_ fixes: #880

## Summary
Continues with @carolinan work in #879 so im borrowing the summary from there. The cover block sets the links to currentColor so that the text and link color has a high contrast against a chosen palette color. This style was overriding the button block and file block link colors. 

Additionally, this PR attempts to address all Blocks that use a button and use the same border-width style consistently. 
## Test instructions

This PR can be tested by following these steps:

1. Create a post with this [sample content](https://cloudup.com/cXTzNhag6EV)
2. Save and view the front
3. Compare the colors and confirm that all the text is readable.
4. Test hover and focus styles.
5. Turn Dark Mode on or off, test both body background colors.
6. Repeat step 5 with the Block editor.

## Screenshots (UPDATED: 2020-11-25)

Before | After | Before (darkmode) | After (darkmode) 
------------|------------|------------|------------
image|image|image|image

## Todo

- [x] Finish solution for nested button colors on a lighter background color.
- [ ] Test and apply this fix to the editor.

## Quality assurance
* [x] I have thought about any security implications this code might add.
* [x] I have checked that this code doesn't impact performance (greatly).
* [x] I have tested this code to the best of my abilities
* [ ] I have checked that this code does not affect the accessibility negatively

### CODE REVIEW

**Comment by carolinan** [comment]
I did not understand why ` !important` is needed for the search block? Or if there was some mismatch with the previous padding?

**Comment by carolinan** [comment]
Also we need to watch the search form (not block) and search widget submit button, it now has no visible border.

**Comment by allancole** [comment]
> I did not understand why !important is needed for the search block? Or if there was some mismatch with the previous padding?

Yeah, I still need to audit much of the changes I’ve made here. There’s quite a few `!important` rules that can probably be removed. 

Additionally, this is likely to turn into quite a large refactor as I work on it more. Ideally we’d want the majority of the button style rules to live in the `button-style()` mixin so we’re sure all buttons follow the same standard behaviors. This makes the block-specific overrides easier to deal with and we’ll hopefullly end up with less CSS lines overall in the end. 

> Also we need to watch the search form (not block) and search widget submit button, it now has no visible border.

Ahh, good catch. Should be easy to attach the widget submit button to the button mixin. Widget form borders should be pretty straightforward too.

**Comment by ryelle** [comment]
Thanks for the really comprehensive test content 💯 

Here are a few issues I caught running through this - I see you're still working on the light-group issues so I left those out.

The "outline" button with a background color has no border (both custom and palette colors):
light-mode-no-bg-link

It's more obvious on focus:
light-mode-no-bg-focus

In the focus states for outline buttons on dark group backgrounds (light site background), only the default button has a white focus, the rest are dark
light-mode-dark-section-focus

> Ideally we’d want the majority of the button style rules to live in the `button-style()` mixin so we’re sure all buttons follow the same standard behaviors.

This sounds great, it would cut down on a lot of these edge cases if it can work out 🙂

**Comment by carolinan** [comment]
The buttons _are_ different though, they do not follow the same behavior, only the styles that are used by all of them should be in the mixin?

**Comment by allancole** [comment]
> The buttons are different though, they do not follow the same behavior, only the styles that are used by all of them should be in the mixin?

@carolinan, If i’m understanding you correctly here, yes but what I’m proposing is we establish the default button style in the `button-style()` mixin as the _single source of truth_. Then, when we need to add tweaks for buttons in different contexts (button block, search form blocks, widgets, comment form etc.), we can do context-specific overrides in each relevant Sass partial. Let me know if that makes sense. The last few commits make for a better example of what I mean where we get something closer to what users expect, but also with less lines of overriding CSS :-)

**Comment by allancole** [comment]
@ryelle, I think those last few commits should solve all three of the issues you found plus fixes the ”light-group“ background issue :-)

**Comment by carolinan** [comment]
404 page search form: Button is missing border when focused.
Widget search form: Button is missing border when focused.

-----
Search block (default body background color)

Default, Button is missing border except on hover. On hover: Black text over grey background is unreadable.
Button inside:  Button is missing border except on hover. On hover: Black text over grey background is unreadable.
Button inside, icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
Button only, with icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
Button only, without icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.

Search block (dark mode enabled)
Button inside: On hover: The minimal color change is not different enough from the default state (accessibility). 
                       On focus: Text and icon are very light grey over white background, the contrast is not high enough (accessibility).

**Comment by carolinan** [comment]
Form submit button, for example the comment form: Border is missing on focus.

**Comment by carolinan** [comment]
On default body background, in the editor:
On button creation, only the dotted outline is visible. 

On hover: Outlined buttons without color settings have grey background and dark grey text.

-----

On dark body backgrounds (not dark mode), in the editor:
On button creation, only the dotted outline is visible. 

The placeholder text is not visible until the button is hovered over.
On hover: Outlined buttons without color settings have white background and white text.

----
With dark mode enabled, in the editor:
On button creation, only the dotted outline is visible. 
The placeholder text is white on white background
On hover: Outlined buttons without color settings have white background and white text.

-----


Because the text is white on white background when hovered, you can't see what you are typing inside the button unless you first move the mouse.

**Comment by carolinan** [comment]
On the default body background:

Buttons placed in a block that has the grey palette color as background are invisible, only the text shows.
**This issue exists with and without this PR.**
On any other light background, the button background turns black and is visible.


default (filled) button on the default background, on hover (yup)
default

With pink (sorry, red) body background, default (filled) button, on hover:
pink

Here, the contrast is too low to be accessible, but low contrast is better than invisible.

**Comment by allancole** [comment]
> 404 page search form: Button is missing border when focused.
> Widget search form: Button is missing border when focused.

Should be fixed in the latest commit :-) 

> **Search block (default body background color)**
> Default, Button is missing border except on hover. On hover: Black text over grey background is unreadable.
> Button inside: Button is missing border except on hover. On hover: Black text over grey background is unreadable.
> Button inside, icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
> Button only, with icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
> Button only, without icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.

Here what I see on this PR for each of those options you listed: 



Could you share a screenshot of what you see @carolinan ? 

> **Search block (dark mode enabled)**
> Button inside: On hover: The minimal color change is not different enough from the default state (accessibility).
> On focus: Text and icon are very light grey over white background, the contrast is not high enough (accessibility).

This I do see ;-). Will try to work out a solution shortly.

**Comment by carolinan** [comment]
Search block:
The border still disappears when I click on the buttons.
The color of the "button inside" is still incorrect when I click the button.

PC, win 10, Chrome version 87.0.4280.66. Gutenberg is not active.


search-click

Compared with trunk:

search-trunk

File: assets/sass/05-blocks/search/_style.scss [hunk 1]
```diff
@@ -52,12 +52,26 @@
...
+		&:active,
 		&:hover {
-			background-color: var(--button--color-background);
-			color: var(--button--color-text);
+			color: var(--button--color-text) !important;
+			background-color: var(--button--color-background) !important;
+			border-color: transparent;
+		}
+
+		&:focus {
```
Comment (carolinan) [inline assets/sass/05-blocks/search/_style.scss:72]: I think that removing this  focus style solves https://github.com/WordPress/twentytwentyone/pull/886#issuecomment-736542156 ?

**Comment by carolinan** [comment]
Front: 
With dark mode enabled, the black link background for focused links affects the filled button. But not the outline button.
darkmode

File: assets/sass/05-blocks/search/_style.scss (hunk 1 above)
Comment (kjellr) [reply assets/sass/05-blocks/search/_style.scss:72]: Should be working better with 27f1fe1 + dbf60c7:

Kapture 2020-12-01 at 09 52 19

**Comment by kjellr** [comment]
The front-end seems to be mostly working in my testing (just using the sample data above). But I'm seeing some really glaring issues in the editor still: 

The outline button doesn't have readable hover text by default:

Kapture 2020-12-01 at 09 55 21

And all of the search buttons appear incorrect: 

Screen Shot 2020-12-01 at 9 54 59 AM

**Comment by carolinan** [comment]
Front:

Dark background, 
Search block with button inside, the button is invisible on focus:

image

Dark mode:
Search block with button inside, the button is almost invisible on focus:
image

**Comment by carolinan** [comment]
Front:
When I **click** on an outlined button, the border is invisible:
image

**Pull request closed without merging**